### PR TITLE
Test block catchup

### PR DESCRIPTION
**Pull request by bowenwang1996**
- Add integration test to test block catchup with network

- Fix bugs in block request and block response

- Remove outdated tests in the network crate

Resolves https://github.com/nearprotocol/nearcore/issues/248

### CODE REVIEW

**Comment by nearmax** [comment]
It would be useful to also test `BlockImporter`, however it seems to be hard to do with the current architecture, because we would need to mock `Runtime` which is hard, because we would need to make a trait and add a type parameter to `BlockImporter` which is also ugly.

Alternatively, we could host `Runtime` on a task and communicate with it using channels. Notice, that there was no need to write mocks through type parameters for consensus, block producer, block importer, rpc, etc in this integration test because we have wrapped them into tasks with channels. Since it looks like we are starting to actively touch the code that depends on the `Runtime` it would be beneficial to move it into a task. In tests we would then simply not create the runtime task and instead would make a simple noop task that fakes the state computation, similarly to how this PR mocks `BlockImporter` with `spawn_simple_block_import_task`.

This is also a common concurrency pattern, https://en.wikipedia.org/wiki/Actor_model (popular in Erlang, Scala, Rust), that allows one to control the access to a resource without `Mutex` or `RwLock`. The way it works, is one would create a `Runtime` object and make the task own it, similarly to how we have implemented `BlockImporter`, `BlockProducer`, and `TxFlowTask`. Other tasks then would communicate with `Runtime` using channels.

Actor model is considered to be an alternative to mutex, and there is an argument on the internet, that they have more-or-less the same performance. Tokio devs also favor the actor model: https://github.com/tokio-rs/tokio-core/issues/167#issuecomment-279175083

CC @ilblackdragon since he also expressed concern with proliferation of templates in our code.

File: node/network/src/protocol.rs [hunk 1]
```diff
@@ -156,16 +156,20 @@ impl<B: SignedBlock, Header: BlockHeader, P: Payload> Protocol<B, Header, P> {
...
+        let best_index = self.chain.best_index();
         let mut next_request_id = 0;
+        let mut block_request = None;
+        let mut request_timestamp = None;
         if status.best_index > best_index {
             let request = message::BlockRequest {
                 id: next_request_id,
-                from: BlockId::Number(best_index),
+                from: BlockId::Number(best_index + 1),
                 to: Some(BlockId::Number(status.best_index)),
```
[the remark this answers is not in the document]
Comment (bowenwang1996) [reply node/network/src/protocol.rs]: I think part of this is unnecessary. When two peers connect, they both send status message to each other, so the one that is behind will naturally request blocks from the other one. Will add hash

File: node/network/tests/integration_test.rs [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+
+    // This is very hacky, but I cannot find a proper way
+    // to stop the tokio event loop.
+    std::mem::drop(handle);
+}
+
+// this test is expensive to run. should be ignored when running
+// all tests
+#[test]
+#[ignore]
```
[the remark this answers is not in the document]
Comment (bowenwang1996) [reply node/network/tests/integration_test.rs]: Did you run the test alone or with other tests?

**Comment by azban** [comment]
you should add  `"files.insertFinalNewline": false` to your VS Code config

**Comment by bowenwang1996** [comment]
It appears that the problem with the second test is that the task is waiting for the lock on network_service2 to drop the node and there is a possible deadlock. Not sure what's the best way to circumvent that. Will open another issue for this.

**Pull request merged**